### PR TITLE
feat: Mass transfer feature

### LINKED ISSUE #10870: Remove `items` code duplication between stores
after #10842

we might need to reuse the same logic in future stores

I'd like a solution that has less code duplication, ideally some utils

stores: `listingCart`, `makeOffer`

pls read [thread](https://github.com/kodadot/nft-gallery/pull/10842#discussion_r1720613685_)

----
can we make some utils so we don't duplicate this code , and can reuse it in future stores :D

```suggestion
    setItem(payload: MakingOfferItem) {
         addItem(this.items, payload)
    },
    updateItem(payload: Partial<MakingOfferItem>) {
          updateItem(this.items, payload, payload.id)
    },
    removeItem(id: ID) {
      removeItem(this.items, id)
    },
```

_Originally posted by @hassnian in https://github.com/kodadot/nft-gallery/pull/10842#discussion_r1720613685_

### LINKED ISSUE #11072: Mass transfer feature
Once in a while, you want to send more NFTs to various addresses or transfer more NFTs to your second wallet. We don't have a way to do it!

Can we please think about having **a mass transfer** feature? 

Where it can be used
- giveaway and competitions
- when transferring more NFTs to another wallet

It will make things faster!

@kodadot/internal-dev, how hard is it to implement that?

### PR DESCRIPTION
**Pull request by hassnian**
## PR Type

- [ ] Bugfix
- [x] Feature
- [x] Refactoring

## Context

mass nft transfer should work

reused a lot of the listing cart code , will create a separate pr after this is merged,  since now `listingCart` has multiple uses will refactor it accordingly , I think that It's going to be more manageable reviewing and testing wise

@JustLuuuu could you test?

- [x] Closes #11072
- [x] Closes #10870

## Needs Design check

- @exezbcz please review

## Screenshot 📸

- [x] My fix has changed **something** on UI; 

CleanShot 2024-10-09 at 15 23 45@2x

### CODE REVIEW

**Comment by vikiival** [comment]
> 'listVisible' is defined but never used. Allowed unused vars must match /^_/u.

**Comment by hassnian** [comment]
mass sending should work

reused a lot of the listing cart code , will create a separate pr after this is merged,  since now `listingCart` has multiple uses will refactor it accordingly , I think that It's going to be more manageable reviewing and testing wise

**Comment by JustLuuuu** [comment]
Just tested. Works fine for me:

I have one question: when I used the generic Polkadot address (5CRxWu1z86Z9o9LGjv2QNXkxKQzRe4e29YYzdnnsQA3qesZb) I got this popup: 
Screenshot 2024-10-11 at 12 44 30

Solution
-

Can we just change the text to say: Please use your DOT address? Or something more user friendly? This was a bit confusing.

Also, why don't we support generic addresses here?

**Comment by hassnian** [comment]
> Just tested. Works fine for me:
> 
> I have one question: when I used the generic Polkadot address (5CRxWu1z86Z9o9LGjv2QNXkxKQzRe4e29YYzdnnsQA3qesZb) I got this popup: Screenshot 2024-10-11 at 12 44 30
> 
> ## Solution
> Can we just change the text to say: Please use your DOT address? Or something more user friendly? This was a bit confusing.
> 
> Also, why don't we support generic addresses here?

fixed 

CleanShot 2024-10-12 at 11 38 47

File: components/items/ItemsGrid/ItemsGridImageTokenEntity.vue [hunk 1]
```diff
@@ -70,7 +70,7 @@
           />
         </template>
 
-        <template v-else-if="listVisible(urlPrefix)">
```
Comment (Jarsen136) [inline components/items/ItemsGrid/ItemsGridImageTokenEntity.vue:73]: why?

File: composables/transaction/transactionSend.ts [hunk 2]
```diff
@@ -49,11 +52,14 @@ function execSendAssetHub(item: ActionSend, api, executeTransaction) {
...
+  const nft = item.nfts[0]
+
   executeTransaction({
-    address: item.collectionId,
+    address: nft.collectionId,
     abi: item.abi as Abi,
     functionName: 'safeTransferFrom',
-    arg: [accountId.value, item.address, item.nftSn],
+    arg: [accountId.value, item.address, nft.sn],
   })
```
Comment (Jarsen136) [inline composables/transaction/transactionSend.ts:63]: should the `batchTransfer` be implemented before merging?

Comment (hassnian) [reply composables/transaction/transactionSend.ts:63]: `batchTransfer`  is not available on `ERC721`

`safeBatchTransferFrom` is added in `ERC1155` 
https://docs.openzeppelin.com/contracts/3.x/erc1155#batch-operations

@vikiival am I missing something ?

Comment (Jarsen136) [reply composables/transaction/transactionSend.ts:63]: Sorry, I didn't make it clear. I mean supporting batch transfer on EVM, no matter which method on ERC.

However, I saw multiple NFT transfers on EVM prevented on the UI side. So, it's not that urgent to implement batch transfer for now.


image

File: components/items/ItemsGrid/ItemsGridImageTokenEntity.vue (hunk 1 above)
Comment (hassnian) [reply components/items/ItemsGrid/ItemsGridImageTokenEntity.vue:73]: item transfer is available for all VMs 

`listVisible` is used inside `ListingCartMini.vue`

CleanShot 2024-10-21 at 15 56 18@2x

**Review by vikiival — APPROVED** [review]
I have tested and WFM

**Review by Jarsen136 — APPROVED** [review]

**Comment by JustLuuuu** [comment]
wfm

**Pull request merged**